src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +789,76 @@ private void PopulateEncLogTableParameters()
             }
         }
 
+        private void PopulateEncLogTableCustomAttributes()
+        {
+            var attributeMap = CreateExistingAttributeMap(_previousGeneration.MetadataReader, _previousGeneration.CustomAttributesAdded, out var lastRow);
+
+            var nextCustomAttributeRow = lastRow + 1;
+
+            foreach (var customAttributeTarget in _customAttributeTargets)
+            {
+                int row;
+                if (attributeMap.TryGetValue(customAttributeTarget, out var queue) &&
+                    queue.Count > 0)
+                {
+                    // If there is already an attribute for this parent, then pull the next available row number
+                    // off the queue and update it
+                    row = queue.Dequeue();
+                }
+                else
+                {
+                    // Otherwise, either this is a new parent that hasn't had an attribute before, or we've run
+                    // out of existing rows to update. Either way, we want to add to the end of the table.
+                    row = nextCustomAttributeRow++;
+
+                    // Keep track of the addition for passing to the next generation
+                    _customAttributesAdded[row] = customAttributeTarget;
+                }
+
+                _customAttributeRows.Add(row);
+                metadata.AddEncLogEntry(
+                    entity: MetadataTokens.CustomAttributeHandle(row),
+                    code: EditAndContinueOperation.Default);
+            }
+        }
+
+        private static Dictionary<EntityHandle, Queue<int>> CreateExistingAttributeMap(MetadataReader metadataReader, IReadOnlyDictionary<int, EntityHandle> customAttributesAdded, out int lastRow)
+        {
+            // For each custom attribute target, we need to keep a list of which row numbers are used in the CustomAttribute table
+            // so that when we emit the new attributes we can overwrite them.
+            var result = new Dictionary<EntityHandle, Queue<int>>(HandleComparer.Default);
+
+            int index = 1;
```

Comment:
>index [](http://example.com/codeflow?start=16&length=5)

I assume this is "row id", correct? I'd use `rowId` as a name since that's the term that's generally used for 1-based index into metadata tables